pkg/issuer/acme/http/pod.go

```diff
@@ -167,6 +167,9 @@ func (s *Solver) buildDefaultPod(ch *cmacme.Challenge) *corev1.Pod {
 			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(ch, challengeGvk)},
 		},
 		Spec: corev1.PodSpec{
+			NodeSelector: map[string]string{
+				"kubernetes.io/os": "linux",
```

Comment:
My only concern with changing the node selector here is that someone else _could_ have built their own images for other platforms and set them to be used using the flag override on the controller, which in turn this change would break.

Perhaps not changing the selector for acmesolver pods makes most sense, and then working on actually having support for Windows? wdyt?

Reply:
If someone's down the path of building their own images, it feels like that's already outside the supported model for updates. (e.g. for *any* change to the acme solving image, should we be rejecting it because any alternative images won't be compatible, so a straight update to the new cert-manager will break things?)

That has made me think about other nodeSelector bits - we also need the right arch etc, but it looks like we're already producing arm images etc.. So providing windows images is in line with that.. I wonder how much difference this needs to look in the Dockerfile etc for the solver and cert-manager, I've not built anything for windows.

I feel that if we're going down that route, then we should be building windows images for all of the components - `-cainjector`, `-webhook`, `-controller` etc. It doesn't feel consistent to require nodeSelector and Linux Nodes for some components but not all. (Are we likely to see people wanting to run this on a windows-only cluster?)

Reply:
Yep absolutely we should be building _all_ component images, not just one :)

> If someone's down the path of building their own images, it feels like that's already outside the supported model for updates

I think the difference here is that the user isn't really making vast changes to the image, rather they are compiling the binary for a different os/arch, which seems to me fairly acceptable and something we should support.

It's also worth noting that if a user *does* run a mixed-OS cluster, they can set the `nodeSelector` via the Issuer resource here: https://github.com/jetstack/cert-manager/blob/26afeceb2b8f7499269c16670bfdd0f884275cd6/pkg/apis/acme/v1/types_issuer.go#L237-L285

> I wonder how much difference this needs to look in the Dockerfile etc for the solver and cert-manager, I've not built anything for windows.

I _think_ it'd be a case of utilising the nanoserver base image, and compiling our go binary for windows. This is currently done in this rule here: https://github.com/jetstack/cert-manager/blob/26afeceb2b8f7499269c16670bfdd0f884275cd6/build/container.bzl#L60-L63 (those specific lines actually point to how the base image for a particular os/arch is selected).

I can't find an existing issue for building windows compatible images, but let's get one opened and try and get this resolved.

In the meantime I'd be in favour of the changes to the *Helm chart* for the main operator components, and then continuing to recommend users specify a `nodeSelector` themselves on Issuer resources otherwise.

Alternatively, given this can all be set by the user anyway, we *could* just not merge this patch and work on adding windows support anyway. What do you think? :)

Reply:
> I think the difference here is that the user isn't really making vast changes to the image, rather they are compiling the binary for a different os/arch, which seems to me fairly acceptable and something we should support.

Fair

> Alternatively, given this can all be set by the user anyway, we could just not merge this patch and work on adding windows support anyway. What do you think? :)

Yeah, it can all be done already, I wanted to improve the default behaviour so those settings aren't necessary. Doing it via providing windows images is just as good. I do feel that we should do this in one way or the other, not change only some of the default nodeSelectors. Happy to not merge this and go the other route :)

Reply:
I've written this up at https://github.com/jetstack/cert-manager/issues/3606

Reply:
```suggestion
				"kubernetes.io/os": runtime.GOOS,
```

_If_ there was any change here I would use the currently running arch, as at least there will be at least one node running that OS!